File: FlashQuizTests/StartQuizViewControllerTests.m
//
//  StartQuizViewControllerTests.m
//  SmallSharpTools
//
//  Created by Brennan Stehling on 7/10/16.
//  Copyright © 2016 FlashQuiz. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "AppDelegate.h"
#import "StartQuizViewController.h"

CG_INLINE UINavigationController *GetNavigationController() {
    AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    return (UINavigationController *)appDelegate.window.rootViewController;
}

@interface StartQuizViewControllerTests : XCTestCase

@end

@implementation StartQuizViewControllerTests

- (void)setUp {
    [super setUp];
}

- (void)tearDown {
    [super tearDown];
}

- (void)testViewControllerGenerically {
    // This is an example of a functional test case.
    // Use XCTAssert and related functions to verify your tests produce the correct results.

//    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
//    XCTAssertNotNil(storyboard);
//
//    UINavigationController *nc = (UINavigationController *)[storyboard instantiateInitialViewController];

    XCTestExpectation *expectation = [self expectationWithDescription:@"UI"];

    UINavigationController *nc = GetNavigationController();
    XCTAssertNotNil(nc);
    XCTAssertTrue([nc isKindOfClass:[UINavigationController class]]);

    [nc popViewControllerAnimated:NO];

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.25 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        StartQuizViewController *vc = (StartQuizViewController *)nc.topViewController;
        XCTAssertNotNil(vc);
        XCTAssertTrue([vc isKindOfClass:[StartQuizViewController class]]);

        // Runs the button tap action.
        [vc performSegueWithIdentifier:@"pushQuizSession" sender:vc];

        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.25 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            [nc popViewControllerAnimated:NO];
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.25 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                [expectation fulfill];
            });
        });
    });

    NSTimeInterval timeout = 60; // 10 seconds
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError * _Nullable error) {
        // do nothing
    }];
}

@end

File: FlashQuizSwiftTests/StartQuizViewControllerTests.swift
//
//  StartQuizViewControllerTests.swift
//  FlashQuiz
//
//  Created by Brennan Stehling on 7/17/16.
//  Copyright © 2016 SmallSharpTools. All rights reserved.
//

import XCTest
import UIKit

class StartQuizViewControllerTests: XCTestCase {

    override func setUp() {
        super.setUp()
    }

    override func tearDown() {
        super.tearDown()
    }

    func testViewControllerGenerically() {
        let expectation = self.expectation(description: "UI")

        guard let nc = getNavigationController() else {
            XCTFail("Failed to get navigation controller")
            return
        }

        nc.popToRootViewController(animated: false)

        guard let vc = nc.topViewController as? StartQuizViewController else {
            XCTFail()
            return
        }

        vc.performSegue(withIdentifier: "pushQuizSession", sender: vc)

        DispatchQueue.main.asyncAfter(deadline: whenInSeconds(1.5)) {
            nc.popToRootViewController(animated: false)
            DispatchQueue.main.asyncAfter(deadline: self.whenInSeconds(0.75)) {
                expectation.fulfill()
            }
        }

        let timeout: TimeInterval = 10
        self.waitForExpectations(timeout: timeout) { (error) in
            // do nothing
        }
    }

    // MARK: Private -

    fileprivate func whenInSeconds(_ seconds: Double) -> DispatchTime {
        return DispatchTime.now() + Double(Int64(seconds * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC)
    }

    fileprivate func getNavigationController() -> UINavigationController? {
        if let appDelegate = UIApplication.shared.delegate as? AppDelegate,
            let navigationController = appDelegate.window.rootViewController as? UINavigationController {
            return navigationController
        }

        return nil
    }

}
